relay-server/src/actors/processor.rs

```diff
@@ -1274,6 +1284,77 @@ impl EnvelopeProcessorService {
         }
     }
 
+    fn handle_submit_envelope(&self, message: SubmitEnvelope) {
+        let SubmitEnvelope { mut envelope } = message;
+
+        #[cfg(feature = "processing")]
+        if self.inner.config.processing_enabled() {
+            if let Some(store_forwarder) = self.inner.store_forwarder.clone() {
+                relay_log::trace!("sending envelope to kafka");
+                store_forwarder.send(StoreEnvelope { envelope });
+                return;
+            }
+        }
+
+        // if we are in capture mode, we stash away the event instead of forwarding it.
+        if Capture::should_capture(&self.inner.config) {
+            relay_log::trace!("capturing envelope in memory");
+            self.inner
+                .test_store
+                .send(Capture::accepted(envelope.take_envelope()));
+            envelope.accept();
+            return;
+        }
+
+        relay_log::trace!("sending envelope to sentry endpoint");
+        let http_encoding = self.inner.config.http_encoding();
+        match encode_envelope(envelope.envelope_mut(), http_encoding) {
+            Ok(body) => {
+                self.inner.upstream_relay.send(SendRequest(SendEnvelope {
+                    envelope,
+                    body,
+                    http_encoding,
+                    project_cache: self.inner.project_cache.clone(),
+                }));
+            }
+            Err(error) => {
+                // Errors are only logged for what we consider an internal discard reason. These
+                // indicate errors in the infrastructure or implementation bugs.
+                relay_log::error!(
+                    error = &error as &dyn Error,
+                    tags.project_key = %envelope.scoping().project_key,
+                    "failed to serialize envelope payload"
+                );
+
+                envelope.reject(Outcome::Invalid(DiscardReason::Internal));
+            }
+        }
+    }
+
+    fn handle_submit_client_reports(&self, message: SubmitClientReports) {
+        let SubmitClientReports {
+            client_reports,
+            scoping,
+        } = message;
+
+        let upstream = self.inner.config.upstream_descriptor();
+        let dsn = PartialDsn::outbound(&scoping, upstream);
+
+        let mut envelope = Envelope::from_request(None, RequestMeta::outbound(dsn));
+        for client_report in client_reports {
+            let mut item = Item::new(ItemType::ClientReport);
+            item.set_payload(ContentType::Json, client_report.serialize().unwrap()); // TODO: unwrap OK?
```

Comment:
re TODO: I would say so, we're serializing our own datastructure and this is already common practice in the rest of Relay.

Reply:
This is copied from the other location, I'd prefer not to change the implementation too much.

Reply:
I just meant we should get rid of the `TODO` and phrased it horribly.